# **Waze Project**
**Course 2 - Get Started with Python**

Welcome to the Waze Project!

Your Waze data analytics team is still in the early stages of their user churn project. Previously, you were asked to complete a project proposal by your supervisor, May Santner. You have received notice that your project proposal has been approved and that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).

A Python notebook has been prepared to guide you through this project. Answer the questions and create an executive summary for the Waze data team.

# **Course 2 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis. This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>

**The purpose** of this project is to investigate and understand the data provided.

**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings.
<br/>

**Part 1:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities

* Compile summary information about the data to inform next steps

**Part 2:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables


<br/>

Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work.



# **Identify data types and compile summary information**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**

Throughout these project notebooks, you'll see references to the problem-solving framework, PACE. The following notebook components are labeled with the respective PACE stages: Plan, Analyze, Construct, and Execute.

<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:

*   `import pandas as pd`

*   `import numpy as np`


In [1]:
# Import packages for data manipulation
import numpy as np
import pandas as pd

Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Task 2b. Summary information**

View and inspect summary information about the dataframe

*Consider the following questions:*
1. When reviewing the `info()` output, what are the data types? How many rows and columns do you have?

2. Does the dataset have any missing values?

In [3]:
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


1: Our datatypes are: **int64, object, ** and **float64**
2: We can see in the 'info()' cell's output that the **label** column is missing 700 values.

### **Task 2c. Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values:

**Question:** Is there a discernible difference between the two populations?


In [13]:
# Isolate rows with null values
mask = df['label'].isna()
print(len(mask[mask == True])) # Double check our mask
df_null = df[mask]
df_null.head()

700


ID label  sessions  drives  total_sessions  n_days_after_onboarding  \
77    77   NaN        63      50      133.104155                      783   
80    80   NaN       116      93      436.060183                     1584   
98    98   NaN        78      64      583.492789                     3414   
111  111   NaN       106     102      113.379056                     2228   
142  142   NaN        32      26      222.129310                      208   

     total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
77                      201                       0       2649.015822   
80                      283                      62       4183.409514   
98                        0                       0       1811.140893   
111                      14                       0       2817.481840   
142                      55                      10       2459.816477   

     duration_minutes_drives  activity_days  driving_days   device  
77               1517.209970             19            13   iPhone  
80               3121.889952             18            15   iPhone  
98                642.189122             12            11  Android  
111              2011.724274             17            13  Android  
142               874.427617             11             7   iPhone

In [17]:
# Display summary stats of rows with null values
df_null.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [21]:
# Isolate rows without null values
df_not_null = df[~mask]
print(df_not_null.shape) # Check

# Display summary stats of rows without null values
df_not_null.describe()

(14299, 13)


ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

All of the metrics seem to be fairly similar purely from a quick glance, but more inspection + visualizations would be necessary to be more certain.

### **Task 2d. Null values - device counts**

Next, check the two populations with respect to the `device` variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [26]:
# Get count of null values by device
android_null = df_null[df_null['device'] == 'Android']
android_null

ID label  sessions  drives  total_sessions  n_days_after_onboarding  \
98        98   NaN        78      64      583.492789                     3414   
111      111   NaN       106     102      113.379056                     2228   
266      266   NaN        70      58      137.151279                     3264   
283      283   NaN       170     137      354.297662                      869   
288      288   NaN       199     159      248.560854                     2959   
...      ...   ...       ...     ...             ...                      ...   
14595  14595   NaN        19      16      208.401594                     2832   
14719  14719   NaN       103     103      122.883011                     2074   
14768  14768   NaN        58      52      119.006357                     2833   
14903  14903   NaN        15      15      112.272365                     1901   
14945  14945   NaN        34      29      134.416604                     1643   

       total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
98                          0                       0       1811.140893   
111                        14                       0       2817.481840   
266                        45                       0       1305.946778   
283                        18                      25       3914.401147   
288                       127                       0       5285.296516   
...                       ...                     ...               ...   
14595                      40                       0       7692.120343   
14719                     109                      14       3966.845788   
14768                      96                      53       1713.734121   
14903                      41                      33       4820.052948   
14945                     268                       2       4554.007843   

       duration_minutes_drives  activity_days  driving_days   device  
98                  642.189122             12            11  Android  
111                2011.724274             17            13  Android  
266                 996.394443             21            18  Android  
283                2500.266993              3             2  Android  
288                1383.325300             29            20  Android  
...                        ...            ...           ...      ...  
14595              4781.852844             17            11  Android  
14719              3121.077111             25            16  Android  
14768              1071.852565             11             7  Android  
14903              2987.498434              7             6  Android  
14945              1579.211201             18            17  Android  

[253 rows x 13 columns]

253 of the 700 entries with a null label value are Android users.

In [27]:
iphone_null = df_null[df_null['device'] == 'iPhone']
iphone_null

ID label  sessions  drives  total_sessions  n_days_after_onboarding  \
77        77   NaN        63      50      133.104155                      783   
80        80   NaN       116      93      436.060183                     1584   
142      142   NaN        32      26      222.129310                      208   
162      162   NaN         3       3       17.250220                     3203   
176      176   NaN       283     226      529.533944                     1110   
...      ...   ...       ...     ...             ...                      ...   
14931  14931   NaN        27      23      130.458842                     2152   
14941  14941   NaN       191     160      485.328204                     1287   
14943  14943   NaN        48      38       96.797017                      555   
14972  14972   NaN       220     181      256.212166                     1718   
14993  14993   NaN        67      57       97.570074                     1131   

       total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
77                        201                       0       2649.015822   
80                        283                      62       4183.409514   
142                        55                      10       2459.816477   
162                        22                       0      11819.366330   
176                         6                       0       3028.601300   
...                       ...                     ...               ...   
14931                       0                       0       2745.719618   
14941                      25                       0       6468.181924   
14943                       0                       6       8266.129497   
14972                     360                      23       5586.913459   
14993                     207                     102       2267.052913   

       duration_minutes_drives  activity_days  driving_days  device  
77                 1517.209970             19            13  iPhone  
80                 3121.889952             18            15  iPhone  
142                 874.427617             11             7  iPhone  
162                3351.019594              3             0  iPhone  
176                2489.697005             17            11  iPhone  
...                        ...            ...           ...     ...  
14931              2035.743475             16            10  iPhone  
14941              3466.104564             14            14  iPhone  
14943              5902.351711             19            19  iPhone  
14972              4104.440202             19            18  iPhone  
14993               318.120634             27            26  iPhone  

[447 rows x 13 columns]

The remaining 447 are iPhone users.

Now, of the rows with null values, calculate the percentage with each device&mdash;Android and iPhone. Use the value_counts() function to achieve this.

In [54]:
# Calculate % of iPhone nulls and Android nulls
print('Android percentage (null): ' + str(100 * android_null['device'].value_counts()[0] / 700) + '%')
print('iPhone percentage (null): ' + str(100 * iphone_null['device'].value_counts()[0] / 700) + '%')

Android percentage (null): 36.142857142857146%
iPhone percentage (null): 63.857142857142854%


How does this compare to the device ratio in the full dataset?

In [68]:
# Calculate % of iPhone users and Android users in full dataset
print('Total Android percentage: ' + str(100 * df[df['device'] == 'Android'].shape[0] / 14999) + '%')
print('Total iPhone percentage: ' + str(100 * df[df['device'] == 'iPhone'].shape[0] / 14999) + '%')

Total Android percentage: 35.515701046736446%
Total iPhone percentage: 64.48429895326355%


The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [80]:
# Create df with churned and retained values for later use + simplicity's sake
df_churned = df[df['label'] == 'churned']
df_retained = df[df['label'] == 'retained']

# Calculate counts of churned vs. retained
num_churned = df_churned.shape[0]
num_retained = df_retained.shape[0]

# Print our results
print('Number churned: ' + str(num_churned))
print('Percent churned (excluding null entries): ' + str(100 * num_churned / 14299))

print('Number retained: ' + str(num_retained))
print('Percent retained (excluding null entries): ' + str(100 * num_retained / 14299))

Number churned: 2536
Percent churned (excluding null entries): 17.73550597943912
Number retained: 11763
Percent retained (excluding null entries): 82.26449402056087


This dataset contains ~82% retained users and ~18% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that we don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [96]:
# Calculate median values of all columns for churned and retained users
print('Churned median values:')
print(df_churned.median())
print('\n')
print('Retained median values:')
print(df_retained.median())

Churned median values:
ID                         7477.500000
sessions                     59.000000
drives                       50.000000
total_sessions              164.339042
n_days_after_onboarding    1321.000000
total_navigations_fav1       84.500000
total_navigations_fav2       11.000000
driven_km_drives           3652.655666
duration_minutes_drives    1607.183785
activity_days                 8.000000
driving_days                  6.000000
dtype: float64


Retained median values:
ID                         7509.000000
sessions                     56.000000
drives                       47.000000
total_sessions              157.586756
n_days_after_onboarding    1843.000000
total_navigations_fav1       68.000000
total_navigations_fav2        9.000000
driven_km_drives           3464.684614
duration_minutes_drives    1458.046141
activity_days                17.000000
driving_days                 14.000000
dtype: float64


This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

Calculate the median kilometers per drive in the last month for both retained and churned users.

In [105]:
# Calculate the medians for each group (retained and churned)
median_km_drives_churned = df_churned['driven_km_drives'].median()
median_km_drives_retained = df_retained['driven_km_drives'].median()

# Divide the median distance by median number of drives
median_km_per_drive_churned = median_km_drives_churned / df_churned['drives'].median()
median_km_per_drive_retained = median_km_drives_retained / df_retained['drives'].median()
print('Churned median km per drive: ' + str(median_km_per_drive_churned))
print('Retained median km per drive: ' + str(median_km_per_drive_retained))

Churned median km per drive: 73.05311332
Retained median km per drive: 73.71669391489361


The median user from both groups drove ~73 km/drive.
Let's calculate how many kilometers per driving day this was!

In [106]:
# Divide the median distance by median number of driving days
median_km_day_churned = median_km_per_drive_churned / df_churned['driving_days'].median()
median_km_day_retained = median_km_per_drive_retained / df_retained['driving_days'].median()
print('Churned: ' + str(median_km_day_churned))
print('Retained: ' + str(median_km_day_retained))

Churned: 12.175518886666666
Retained: 5.265478136778115


Now, we'll calculate the median number of drives per driving day for each group.

In [108]:
# Divide the median number of drives by median number of driving days
drives_per_driving_day_churned = df_churned['drives'].median() / df_churned['driving_days'].median()
drives_per_driving_day_retained = df_retained['drives'].median() / df_retained['driving_days'].median()
print('Churned: ' + str(drives_per_driving_day_churned))
print('Retained: ' + str(drives_per_driving_day_retained))

Churned: 8.333333333333334
Retained: 3.357142857142857


The median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, we'll examine whether there is an imbalance in how many users churned by device type.

Begin by getting the overall counts of each device type for each group, churned and retained.

In [123]:
# For each label, calculate the number of Android users and iPhone users

# Create a dataframe without the rows containing null label values
df_no_null = df[~df['label'].isna()]
df_no_null.shape # verify

num_android = df_no_null[df_no_null['device'] == 'Android'].shape[0]
num_iphone = df_no_null[df_no_null['device'] == 'iPhone'].shape[0]
print(num_android)
print(num_iphone)

5074
9225


Now, within each group, churned and retained, we calculate what percent was Android and what percent was iPhone.

In [131]:
# For each label, calculate the percentage of Android users and iPhone users

# Number of each category's Android users divided by total churned users
churned_android_pct = 100 * df_churned[df_churned['device'] == 'Android'].shape[0] / df_churned.shape[0]
retained_android_pct = 100 * df_retained[df_retained['device'] == 'Android'].shape[0] / df_retained.shape[0]
# Same process for iPhone users
churned_iphone_pct = 100 * df_churned[df_churned['device'] == 'iPhone'].shape[0] / df_churned.shape[0]
retained_iphone_pct = 100 * df_retained[df_retained['device'] == 'iPhone'].shape[0] / df_retained.shape[0]

print("% Churned Android: " + str(churned_android_pct))
print("% Churned iPhone: " + str(churned_iphone_pct))
print("% Retained Android: " + str(retained_android_pct))
print("% Retained iPhone: " + str(retained_iphone_pct))

% Churned Android: 35.13406940063091
% Churned iPhone: 64.86593059936908
% Retained Android: 35.5606562951628
% Retained iPhone: 64.4393437048372


The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.